# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionCtx.h
# Commit: 786ea95e928b
# Full Hash: 786ea95e928b7bcec53a56b50f1d0ff136c76a62
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - Move NoTrialsConfig into SharedWebrtcState and make WebrtcCallWrapper keep a ref to it. r=ng
#   
#   SharedWebrtcState when managed by PeerConnectionCtx is destroyed on main thread.
#   With interactions with call instances being async, there is no guarantee that
#   the objects owned by SharedWebrtcState (and referenced through raw pointers by
# ==============================================================================

diff -r 2a373d864092 -r 786ea95e928b dom/media/webrtc/jsapi/PeerConnectionCtx.h
--- a/dom/media/webrtc/jsapi/PeerConnectionCtx.h	Wed Apr 14 17:40:24 2021 +0200
+++ b/dom/media/webrtc/jsapi/PeerConnectionCtx.h	Fri Apr 16 14:44:09 2021 +0200
@@ -41,7 +41,8 @@
 
   SharedWebrtcState(RefPtr<AbstractThread> aCallWorkerThread,
                     webrtc::AudioState::Config&& aAudioStateConfig,
-                    RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory);
+                    RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory,
+                    UniquePtr<webrtc::WebRtcKeyValueConfig> aTrials);
 
   webrtc::SharedModuleThread* GetModuleThread();
 
@@ -59,6 +60,10 @@
   // instances in large calls.
   const RefPtr<webrtc::AudioDecoderFactory> mAudioDecoderFactory;
 
+  // Trials instance shared between calls, to limit the number of instances in
+  // large calls.
+  const UniquePtr<webrtc::WebRtcKeyValueConfig> mTrials;
+
  private:
   virtual ~SharedWebrtcState();
 
@@ -101,8 +106,6 @@
 
   SharedWebrtcState* GetSharedWebrtcState() const;
 
-  webrtc::WebRtcKeyValueConfig* GetTrials() const { return mTrials.get(); }
-
   // WebrtcGlobalInformation uses this; we put it here so we don't need to
   // create another shutdown observer class.
   mozilla::dom::Sequence<mozilla::dom::RTCStatsReportInternal>
@@ -155,14 +158,10 @@
 
   // State used by libwebrtc that needs to be shared across all PeerConnections
   // and all Call instances. Set while there is at least one peer connection
-  // registered.
+  // registered. CallWrappers can hold a ref to this object to be sure members
+  // are alive long enough.
   RefPtr<SharedWebrtcState> mSharedWebrtcState;
 
-  // Trials instance shared between calls, to limit the number of instances in
-  // large calls. Note that this has to outlive the Call instances owned by
-  // PeerConnections, and as such is not unset prior to the ctx dtor.
-  UniquePtr<webrtc::WebRtcKeyValueConfig> mTrials;
-
   static PeerConnectionCtx* gInstance;
 
  public: